### PR TITLE
Laboratory work #1, Viktoriya Aleksina - 23FPL2 

### PR DESCRIPTION
**Pull request by AleksinaV**
print hello

### CODE REVIEW

File: seminars/practice_1_intro_strings.py [hunk 1]
```diff
@@ -6,6 +6,12 @@
 
 # pylint: disable=invalid-name
 
+a = b = 17
```
Comment (artyomtugaryov) [inline seminars/practice_1_intro_strings.py:9]: please, revert this file

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Unfortunately, there is not implementation of the first lab. Hope you'll done on time

File: seminars/practice_1_intro_strings.py (hunk 1 above)
Comment (AleksinaV) [reply seminars/practice_1_intro_strings.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -20,6 +24,14 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if type(text) is not str or len(text) == 0:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:27]: use isinstance instead of `type(...) == ..`

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -34,6 +46,14 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if type(tokens) is not list or len(tokens) == 0 or None in tokens:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:49]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -49,6 +69,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
 
     In case of corrupt input arguments, None is returned
     """
+    if type(language) is not str or type(text) is not str or len(language) == 0 or len(text) == 0:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:72]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -64,11 +89,31 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+            or len(predicted) != len(actual)):
+        return None
+
+    value = 0
+    i = 0
+    while i < len(actual):
+        value += (actual[i] - predicted[i]) ** 2
+        i += 1
+
+    mse = value / i
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:102]: return as is without creating the variable

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -64,11 +89,31 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+    sorted_dict = {}
+    for key, value in sorted_list:
+        sorted_dict[key] = value
+    return sorted_dict
 
 
 def compare_profiles(
-    unknown_profile: dict[str, str | dict[str, float]],
-    profile_to_compare: dict[str, str | dict[str, float]],
+        unknown_profile: dict[str, str | dict[str, float]],
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:115]: revert these lines

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -84,12 +129,31 @@ def compare_profiles(
     In case of corrupt input arguments or lack of keys 'name' and
     'freq' in arguments, None is returned
     """
+    if ((type(unknown_profile) is not dict or type(profile_to_compare) is not dict or len(unknown_profile) == 0
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:132]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -84,12 +129,31 @@ def compare_profiles(
...
+    sorted_unknown_profile = list(dict(sorted(unknown_profile_freq.items(), key=lambda x: x[0])).values())
+
+    return calculate_mse(sorted_profile_to_compare, sorted_unknown_profile)
 
 
 def detect_language(
-    unknown_profile: dict[str, str | dict[str, float]],
-    profile_1: dict[str, str | dict[str, float]],
-    profile_2: dict[str, str | dict[str, float]],
+        unknown_profile: dict[str, str | dict[str, float]],
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:154]: revert these lines

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -119,6 +196,12 @@ def load_profile(path_to_file: str) -> dict | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if type(path_to_file) is not str or len(path_to_file) == 0:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:199]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -149,10 +264,20 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
+        preprocessed_profile = preprocess_profile(load_profile(path))
+        if preprocessed_profile is not None:
+            profiles_list.append(preprocessed_profile)
+
+    return profiles_list
 
 
 def detect_language_advanced(
-    unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
+        unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:280]: revert

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -167,6 +292,19 @@ def detect_language_advanced(
...
     """
+    if (type(unknown_profile) is not dict or type(known_profiles) is not list or len(unknown_profile) == 0
+            or len(known_profiles) == 0):
+        return None
+
+    result_list = []
+    for profile in known_profiles:
+        result_list.append((profile['name'], compare_profiles(unknown_profile, profile)))
+    result_list = sorted(sorted(result_list, key=lambda x: x[0]), key=lambda x: x[1], reverse=False)
+    # for profile in result_list:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:303]: remove

File: lab_1_classify_by_unigrams/start.py [hunk 12]
```diff
@@ -1,20 +1,67 @@
...
     """
+
     with open("assets/texts/en.txt", "r", encoding="utf-8") as file_to_read_en:
         en_text = file_to_read_en.read()
     with open("assets/texts/de.txt", "r", encoding="utf-8") as file_to_read_de:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:22]: ?

Comment (artyomtugaryov) [reply lab_1_classify_by_unigrams/start.py:22]: remove commented code

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/start.py (hunk 12 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/start.py:22]: done

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -20,6 +21,14 @@ def tokenize(text: str) -> list[str] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(text, str) or len(text) == 0:
+        return None
+
+    tokens = []
+    for symbol in text:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:28]: Call lower only once for the source text

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -34,6 +43,14 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
 
     In case of corrupt input arguments, None is returned
     """
+    if not isinstance(tokens, list) or len(tokens) == 0 or None in tokens:
+        return None
+
+    freq_dict = {}
+    for character in tokens:
+        freq_dict[character] = (tokens.count(character) / len(tokens))
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:51]: You call `list.count`  for each element from the list. What if there are some identical elements and you get the same results for this elements but could run it only for unique letters. Can you optimize it?

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -64,6 +89,18 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
     In case of corrupt input arguments, None is returned
     """
+    if (not isinstance(predicted, list) or not isinstance(actual, list)
+            or len(predicted) == 0 or len(actual) == 0
+            or len(predicted) != len(actual)):
+        return None
+
+    value = 0
+    i = 0
+    while i < len(actual):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:99]: Woow.... You don't like `for`, right?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Perfect! Need only small fixes

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py:53]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done (I hope)

File: lab_1_classify_by_unigrams/main.py [hunk 16]
```diff
@@ -135,6 +221,38 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+            profile_freq_key = profile_freq[key] / profile['n_words'][0]
+            profile_freq_key_upper = profile_freq[key.upper()] / profile['n_words'][0]
+            result_freq.update({key.lower(): profile_freq_key + profile_freq_key_upper})
+            del profile_freq[key]
+            del profile_freq[key.upper()]
+        else:
+            profile_freq_key = profile_freq[key] / profile['n_words'][0]
+            result_freq.update({key.lower(): profile_freq_key})
+
+    del profile['n_words']
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:250]: do not use `del`!

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (AleksinaV) [reply lab_1_classify_by_unigrams/main.py]: done

**Pull request closed without merging**